### PR TITLE
Store End Device CAC in the IS

### LINKED ISSUE #5436: Store End Device Claim Authentication Code in the IS
#### Summary

Store End Device Claim Authentication Code in the IS

#### Why do we need this?

References a discussion on Slack; https://thethingsindustries.slack.com/archives/C8SCJ2GG1/p1649926239060559
 We currently store the End Device CAC only on the JS and this flow isn't available with external JSes.
We need to store this somewhere so that it can be used to generate QR Codes. The most logical place would be the IS.

#### What is already there? What do you see now?

The CAC is not stored anywhere in our new claiming flow.

#### What is missing? What do you want to see?

Store the CAC in the IS as part of the device proto.

#### Environment

TTS v3.19.1

#### How do you propose to implement this?

This is mostly a duplicate of the Gateway CAC storage implementation in the IS.

- We need a migration script to fetch existing devices in the JS registry that has the CAC and store it in the IS to make things consistent.

#### How do you propose to test this?

Local stack and UT

#### Can you do this yourself and submit a Pull Request?

Yes.

cc: @johanstokking @nejraselimovic

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #5436 

#### Changes

- Store End Device CAC in the IS
- Update CLI

#### Testing

- [x] Remove migration script
- [x] UT
- [x] E2E Test

```
----
CRUD
----

1. Create a new device with a CAC

./ttn-lw-cli dev create --application-id test-app < ~/Documents/dev.json
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
INFO  Device not configured for claiming, register in the cluster Join Server
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T07:40:08.285Z",
  "updated_at": "2022-07-20T07:40:08.760313Z",
  "name": "My Device",
  "description": "Living room temperature sensor",
  "attributes": {},
  "network_server_address": "localhost",
  "application_server_address": "localhost",
  "join_server_address": "localhost",
  "lorawan_version": "MAC_V1_0_2",
  "lorawan_phy_version": "PHY_V1_0_2_REV_B",
  "frequency_plan_id": "EU_863_870_TTN",
  "supports_join": true,
  "claim_authentication_code": {
    "value": "SECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

2. Read

$./ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T07:40:08.285Z",
  "updated_at": "2022-07-20T07:40:08.285Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "SECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

3. Update

$  ./ttn-lw-cli dev set test-app test-dev --claim-authentication-code.valid_from 2022-05-02T00:00:00Z --claim-authentication-code.valid_to 2023-05-09T00:00:00Z --claim-authentication-code.value SECRET2
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T07:40:08.285Z",
  "updated_at": "2022-07-20T07:40:37.159Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "SECRET2",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

4. Get after update

./ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T07:40:08.285Z",
  "updated_at": "2022-07-20T07:40:37.159Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "SECRET2",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

5. Delete the device

/ttn-lw-cli dev delete test-app test-dev
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
```

- [x] Regression

```
----
Backwards Compatibility Check
----
1. Set using older version (Warning log printed)

$ ttn-lw-cli version
The Things Network Command-line Interface: ttn-lw-cli
Version:             3.20.2
Build date:          2022-07-20T07:58:54Z
Git commit:          65556a319
Go version:          go1.18.4
OS/Arch:             darwin/amd64


$ ttn-lw-cli dev set test-app test-dev --claim-authentication-code.valid_from 2022-05-02T00:00:00Z --claim-authentication-code.valid_to 2023-05-09T00:00:00Z --claim-authentication-code.value SECRET7
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T10:06:49.819243Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "SECRET7",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}
INFO  New patch version available {"current": "3.20.1-dev", "docs_url": "https://www.thethingsindustries.com/docs/getting-started/upgrading/", "latest": "3.20.2"}

WARN  Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead  {"grpc.method": "Set", "grpc.service": "ttn.lorawan.v3.JsEndDeviceRegistry", "namespace": "grpc", "request_id": "01G8DB2A1V5EG7WRH46F5A6DG9"}

2. Get using new version (with warning)

./ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
WARN  Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T08:15:16.392029Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "JSSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

3. Get using older version (Read from Join Server)

ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T10:06:49.819243Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "JSSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

3. Set IS from new CLI and get from same CLI

./ttn-lw-cli dev set test-app test-dev --claim-authentication-code.valid_from 2022-05-02T00:00:00Z --claim-authentication-code.valid_to 2023-05-09T00:00:00Z --claim-authentication-code.value ISSECRET
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T10:14:27.763Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "ISSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

./ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T10:14:27.763Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "ISSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}

4. Get using older CLI

ttn-lw-cli dev get test-app test-dev --claim-authentication-code
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T08:01:05.828Z",
  "updated_at": "2022-07-20T10:14:27.763Z",
  "join_server_address": "localhost",
  "claim_authentication_code": {
    "value": "JSSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
}
```

- [x] QR Code generator check

```
1. Create older CLI (uses JSSECRET)
ttn-lw-cli dev generate-qr test-app test-dev --format-id tr005
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
INFO  Generated QR code {"filename": "/Users/krishnaiyereaswaran/test-dev.png", "value": "LW:D0:0000000000000000:1122334455667788:00000000:OJSSECRET"}
INFO  New patch version available {"current": "3.20.1-dev", "docs_url": "https://www.thethingsindustries.com/docs/getting-started/upgrading/", "latest": "3.20.2"}

2. Create using new CLI (uses ISSecret)
/ttn-lw-cli dev generate-qr test-app test-dev --format-id tr005
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
INFO  Generated QR code {"filename": "/Users/krishnaiyereaswaran/go/src/go.thethings.network/lorawan-stack/test-dev.png", "value": "LW:D0:0000000000000000:1122334455667788:00000000:OISSECRET"}
INFO  New patch version available {"current": "3.20.1-dev", "docs_url": "https://www.thethingsindustries.com/docs/getting-started/upgrading/", "latest": "3.20.2"}

3. Clear IS Secret and try again with new CLI (with warning)
./ttn-lw-cli dev generate-qr test-app test-dev --format-id tr005
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
WARN  Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead
INFO  Generated QR code {"filename": "/Users/krishnaiyereaswaran/go/src/go.thethings.network/lorawan-stack/test-dev.png", "value": "LW:D0:0000000000000000:1122334455667788:00000000:OJSSECRET"}
INFO  New patch version available {"current": "3.20.1-dev", "docs_url": "https://www.thethingsindustries.com/docs/getting-started/upgrading/", "latest": "3.20.2"}
```

- [x] Claim APP

The Claim App (part of enterprise) requires [a single change](https://github.com/TheThingsIndustries/lorawan-stack/commit/166e3bb18b25af729bd45900e0028692b5c4d1d7) to work with these changes. Here are the tests

Claim via QR Code
Screenshot 2022-07-20 at 18 12 06

Claiming with an invalid QR Code fails
```
./tti-lw-cli dev set test-claimed-app test-dev --claim-authentication-code.valid_from 2022-05-02T00:00:00Z --claim-authentication-code.valid_to 2023-05-09T00:00:00Z --claim-authentication-code.value INVALIDSECRET
WARN  Using insecure connection to OAuth server
WARN  Using insecure connection to API
{
  "ids": {
    "device_id": "test-dev",
    "application_ids": {
      "application_id": "test-claimed-app"
    },
    "dev_eui": "1122334455667788",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-07-20T16:11:59.891Z",
  "updated_at": "2022-07-20T16:14:09.540Z",
  "join_server_address": "thethings.localhost",
  "claim_authentication_code": {
    "value": "INVALIDSECRET",
    "valid_from": "2022-05-02T00:00:00Z",
    "valid_to": "2023-05-09T00:00:00Z"
  }
```
Screenshot 2022-07-20 at 18 14 53

##### Regressions

Possible regressions have been tested for. Please see tests above.

#### Notes for Reviewers

The migration script from the original implementation has been removed.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Review by johanstokking — COMMENTED** [review]
@KrishnaIyer

- This should target `v3.21` branch. Is there something not finished that this is draft?
- Please remove the import/export commands, this is covered [here](https://github.com/TheThingsIndustries/infrastructure/issues/973#issuecomment-1174395306) with temporary tooling
- What do we do with the Join Server's claim authentication code? Should we deprecate the field? Without breaking API, how do we ensure that the IS's CAC is used and not JS's? For example when generating QR codes

@htdvisser please review IS part

**Comment by johanstokking** [comment]
As discussed offline; we'll keep the same field, clients will fetch it from JS and IS (with IS taking precedence) and only store it in IS. JS will return a warning when setting the claim authentication code. We can't stop clients from doing that for API compatibility, so there should be a `bump/major` issue to get rid of that.

**Comment by KrishnaIyer** [comment]
This PR is ready now.
- [x] Implementation
- [x] Extensive testing (See description)
- [x] Check for regressions and supported functions (See description)
- [x] Enterprise updates  (See description)
- [x] Issue for removing CAC field in the JS https://github.com/TheThingsNetwork/lorawan-stack/issues/5631
- [x] Console updates to use the IS value

File: pkg/identityserver/end_device_registry.go [hunk 1]
```diff
@@ -188,13 +227,33 @@ func (is *IdentityServer) getEndDevice(ctx context.Context, req *ttnpb.GetEndDev
...
+	}
+
 	err = is.store.Transact(ctx, func(ctx context.Context, st store.Store) (err error) {
 		dev, err = st.GetEndDevice(ctx, req.EndDeviceIds, req.FieldMask.GetPaths())
 		return err
 	})
 	if err != nil {
 		return nil, err
 	}
+	if dev.ClaimAuthenticationCode != nil && dev.ClaimAuthenticationCode.Value != "" {
```
Comment (nicholaspcr) [inline pkg/identityserver/end_device_registry.go]: nit: `if dev.GetClaimAuthenticationCode().GetValue() != "" {`

**Review by nicholaspcr — APPROVED** [review]
Besides the small `nit` and the `data/lorawan-devices` change, the IS section LGTM.

File: pkg/identityserver/end_device_registry.go (hunk 1 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go]: The `GetXXX` method is generally used if we are unsure of the field being nil. In this case `dev.ClaimAuthenticationCode != nil` already checks this and `value` is a string.

Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go]: ah sorry. I guess you mean `dev.GetClaimAuthenticationCode().GetValue()` as a shorthand notation. Indeed that's a good one.

File: api/end_device.proto [hunk 2]
```diff
@@ -740,7 +740,10 @@ message EndDevice {
   // Indicates whether this device represents a multicast group.
   bool multicast = 45;
 
-  // Authentication code to claim ownership of the end device. Stored in Join Server.
+  // Authentication code to claim ownership of the end device.
+  // From TTS v3.21.0 this field is stored in the Identity Server.
+  // For TTS versions < 3.21.0, this field is stored in the Join Server.
+  // The value stored on the Identity Server gets preference.
```
Comment (johanstokking) [inline api/end_device.proto]:
```suggestion
  // The value stored on the Identity Server takes precedence.
```

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 3]
```diff
@@ -319,9 +326,16 @@ var (
...
 			}
-			if len(jsPaths) > 0 && jsMismatch {
-				logger.WithField("paths", jsPaths).Warn("Deselecting Join Server paths")
-				jsPaths = nil
+
+			if len(jsPaths) > 0 {
+				if device.ClaimAuthenticationCode != nil && device.ClaimAuthenticationCode.Value != "" {
+					// ClaimAuthenticationCode is already retrieved from the IS. We can unset the related JS paths.
+					jsPaths = ttnpb.ExcludeFields(jsPaths, claimAuthenticationCodePaths...)
+				}
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: This should be before the `len(jsPaths)` and `jsMismatch` check. Now there will be a warning if the JS paths only contain the CAC fields

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 4]
```diff
@@ -621,10 +642,16 @@ var (
 				}
 			}
 
-			// Require EUIs for devices that need to be added to the Join Server.
-			if len(jsPaths) > 0 && (device.Ids.JoinEui == nil || device.Ids.DevEui == nil) {
-				return errNoEndDeviceEUI.New()
+			if len(jsPaths) > 0 {
+				// Remove Claim Authentication Code related paths from the call to the JS registry.
+				jsPaths = ttnpb.ExcludeFields(jsPaths, claimAuthenticationCodePaths...)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: 1. Shouldn't this be after `splitEndDeviceSetPaths()` or even better, part of that function?
2. Shouldn't there be a TODO comment with the `bump/major` issue to remove this line?

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 5]
```diff
@@ -682,6 +709,10 @@ var (
 				paths = append(paths, "locations")
 			}
 
+			if hasUpdateDeviceClaimAuthenticationCodeFlags(cmd.Flags()) {
+				paths = append(paths, "claim_authentication_code")
+			}
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Why wouldn't this be in `paths` already?

File: pkg/identityserver/config.go [hunk 6]
```diff
@@ -92,6 +92,9 @@ type Config struct {
 		SMTP         smtp.Config          `name:"smtp"`
 		Templates    emailTemplatesConfig `name:"templates"`
 	} `name:"email"`
+	EndDevices struct {
+		EncryptionKeyID string `name:"encryption-key-id" description:"ID of the key used to encrypt end device secrets"`
```
Comment (johanstokking) [inline pkg/identityserver/config.go]:
```suggestion
		EncryptionKeyID string `name:"encryption-key-id" description:"ID of the key used to encrypt end device secrets at rest"`
```

File: pkg/identityserver/end_device_registry.go [hunk 7]
```diff
@@ -151,6 +156,37 @@ func (is *IdentityServer) createEndDevice(ctx context.Context, req *ttnpb.Create
 	}
 	defer func() { is.setFullEndDevicePictureURL(ctx, dev) }()
 
+	// Store plaintext value to return in the response to clients.
+	var ptCACSecret string
+
+	if req.EndDevice.ClaimAuthenticationCode != nil {
+		ptCACSecret = req.EndDevice.ClaimAuthenticationCode.Value
```
Comment (johanstokking) [inline pkg/identityserver/end_device_registry.go:163]: We don't forbid `:` in the value, right? What happens if people put `:` in the CAC and we try to read it?

File: pkg/joinserver/grpc_deviceregistry.go [hunk 8]
```diff
@@ -191,6 +191,17 @@ func (srv jsEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndD
...
 
+	if ttnpb.HasAnyField(
+		req.FieldMask.GetPaths(),
+		"claim_authentication_code.value",
+		"claim_authentication_code.valid_from",
+		"claim_authentication_code.value_to",
+	) {
+		log.FromContext(ctx).Warn(
+			`Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead`, //nolint:lll
+		)
```
Comment (johanstokking) [inline pkg/joinserver/grpc_deviceregistry.go]: This is not a useful warning for the JS operator.

The warning should be returned to the client via response warnings.

**Review by adriansmares — COMMENTED** [review]
This generally looks good to me, but keep in mind that you will also have to implement this functionality in `bunstore` - otherwise this functionality cannot be used while the `bunstore` flag is enabled.

References https://github.com/TheThingsIndustries/lorawan-stack/pull/3314

File: pkg/joinserver/grpc_deviceregistry.go (hunk 8 above)
Comment (KrishnaIyer) [reply pkg/joinserver/grpc_deviceregistry.go]: `response warnings` as in gRPC response?

File: pkg/identityserver/end_device_registry.go (hunk 7 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go:163]: We do forbid all special characters for the value; `^[A-Z0-9]{1,32}$`. 
Speaking of which I think we should also allow lowercase alphabets here right?

File: pkg/joinserver/grpc_deviceregistry.go (hunk 8 above)
Comment (adriansmares) [reply pkg/joinserver/grpc_deviceregistry.go]: Check out `warning.Add`.

File: cmd/ttn-lw-cli/commands/end_devices.go (hunk 5 above)
Comment (KrishnaIyer) [reply cmd/ttn-lw-cli/commands/end_devices.go]: The sub-field will be in the path and the top level will not. However this is better handled in the IS. I've updated this.

**Comment by KrishnaIyer** [comment]
Added Bun implementation; https://github.com/TheThingsIndustries/lorawan-stack/pull/3336

File: pkg/identityserver/end_device_registry.go (hunk 7 above)
Comment (johanstokking) [reply pkg/identityserver/end_device_registry.go:163]: Ah right. Yes, I think we should allow lowercase too.

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 9]
```diff
@@ -319,9 +326,16 @@ var (
...
+
+			if len(jsPaths) > 0 {
+				if device.ClaimAuthenticationCode.GetValue() != "" {
+					// ClaimAuthenticationCode is already retrieved from the IS. We can unset the related JS paths.
+					jsPaths = ttnpb.ExcludeFields(jsPaths, claimAuthenticationCodePaths...)
+				}
+				if jsMismatch {
+					logger.WithField("paths", jsPaths).Warn("Deselecting Join Server paths")
+					jsPaths = nil
+				}
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: This can still be empty because of the `ExcludeFlags` above. Why not move this after L318?

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 10]
```diff
@@ -342,6 +356,13 @@ var (
 				device.UpdatedAt = res.UpdatedAt
 			}
 
+			if device.ClaimAuthenticationCode != nil &&
+				device.ClaimAuthenticationCode.Value != "" &&
+				ttnpb.HasAnyField(jsPaths, claimAuthenticationCodePaths...) {
+				logger.Warn(
+					`Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead`, //nolint:lll
+				)
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Is the user now getting two warnings?

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 11]
```diff
@@ -621,10 +642,13 @@ var (
 				}
 			}
 
-			// Require EUIs for devices that need to be added to the Join Server.
-			if len(jsPaths) > 0 && (device.Ids.JoinEui == nil || device.Ids.DevEui == nil) {
-				return errNoEndDeviceEUI.New()
+			if len(jsPaths) > 0 {
+				// Require EUIs for devices that need to be added to the Join Server.
+				if device.Ids.JoinEui == nil || device.Ids.DevEui == nil {
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: nit; revert

File: cmd/ttn-lw-cli/commands/end_devices.go [hunk 12]
```diff
@@ -1255,6 +1282,13 @@ This command may take end device identifiers from stdin.`,
 				return err
 			}
 
+			if device.ClaimAuthenticationCode.GetValue() != "" &&
+				ttnpb.HasAnyField(jsPaths, claimAuthenticationCodePaths...) {
+				logger.Warn(
+					`Storage of claim authentication code in the Join Server registry is deprecated. Use the Identity Server registry instead`, //nolint:lll
```
Comment (johanstokking) [inline cmd/ttn-lw-cli/commands/end_devices.go]: Idem

File: pkg/identityserver/end_device_registry.go [hunk 13]
```diff
@@ -188,13 +227,33 @@ func (is *IdentityServer) getEndDevice(ctx context.Context, req *ttnpb.GetEndDev
 		defer func() { is.setFullEndDevicePictureURL(ctx, dev) }()
 	}
 
+	if ttnpb.HasAnyField(ttnpb.TopLevelFields(req.FieldMask.GetPaths()), "claim_authentication_code") {
```
Comment (johanstokking) [inline pkg/identityserver/end_device_registry.go:232]: I think `HasAnyField` already considers top level fields, doesn't it?

File: pkg/identityserver/end_device_registry.go [hunk 14]
```diff
@@ -188,13 +227,33 @@ func (is *IdentityServer) getEndDevice(ctx context.Context, req *ttnpb.GetEndDev
...
+		req.FieldMask.Paths = append(req.FieldMask.GetPaths(), "claim_authentication_code")
```
Comment (johanstokking) [inline pkg/identityserver/end_device_registry.go]: There's utility to ensure paths to be included.

What is this trying to do? If only "valid to" is set, to also set value and valid from even if they're empty?

File: pkg/identityserver/end_device_registry.go [hunk 15]
```diff
@@ -283,6 +342,10 @@ func (is *IdentityServer) updateEndDevice(ctx context.Context, req *ttnpb.Update
 		req.FieldMask = cleanFieldMaskPaths(ttnpb.EndDeviceFieldPathsNested, req.FieldMask, nil, []string{"activated_at"})
 	}
 
+	if ttnpb.HasAnyField(ttnpb.TopLevelFields(req.FieldMask.GetPaths()), "claim_authentication_code") {
+		req.FieldMask.Paths = append(req.FieldMask.GetPaths(), "claim_authentication_code")
+	}
```
Comment (johanstokking) [inline pkg/identityserver/end_device_registry.go]: Idem; what is this for?

File: pkg/identityserver/end_device_registry.go [hunk 16]
```diff
@@ -151,6 +156,37 @@ func (is *IdentityServer) createEndDevice(ctx context.Context, req *ttnpb.Create
...
+			// Store the encrypted value along with the ID of the key used to encrypt it.
+			req.EndDevice.ClaimAuthenticationCode.Value = fmt.Sprintf(
+				"%s%s%s",
+				is.config.EndDevices.EncryptionKeyID,
+				endDeviceAuthenticationCodeSeparator,
+				hex.EncodeToString(encrypted),
+			)
+		} else {
+			log.FromContext(ctx).Debug(
+				"No encryption key defined, store end device claim authentication code directly in plaintext",
```
Comment (kschiffer) [inline pkg/identityserver/end_device_registry.go:185]:
```suggestion
				"No encryption key defined, storing end device claim authentication code directly in plaintext",
```
Also do we really want to have this footgun?

File: pkg/identityserver/end_device_registry.go [hunk 17]
```diff
@@ -295,6 +358,40 @@ func (is *IdentityServer) updateEndDevice(ctx context.Context, req *ttnpb.Update
...
```
Comment (kschiffer) [inline pkg/identityserver/end_device_registry.go:392]:
```suggestion
				"No encryption key defined, storing end device claim authentication code directly in plaintext",
```

File: pkg/identityserver/end_device_registry.go [hunk 18]
```diff
@@ -336,6 +438,16 @@ func (is *IdentityServer) deleteEndDevice(ctx context.Context, ids *ttnpb.EndDev
 	return ttnpb.Empty, nil
 }
 
+func validateEndDeviceAuthenticationCode(authCode ttnpb.EndDeviceAuthenticationCode) error {
+	if validFrom, validTo := ttnpb.StdTime(authCode.ValidFrom), ttnpb.StdTime(authCode.ValidTo); validFrom != nil &&
+		validTo != nil {
```
Comment (kschiffer) [inline pkg/identityserver/end_device_registry.go:445]: Shouldn't this also check whether `validTo` is not in the past?

File: pkg/identityserver/end_device_registry.go (hunk 16 above)
Comment (johanstokking) [reply pkg/identityserver/end_device_registry.go:185]: We use imperative tense in log messages

File: pkg/identityserver/end_device_registry.go (hunk 18 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go:445]: One of the ways of making the CAC invalid for testing is to play around with `validFrom` and `validTo`. I get your point but I don't know if we should check for this. 
Users must be aware of what `validTo` means right?

File: pkg/identityserver/end_device_registry.go (hunk 14 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go:231]: https://github.com/TheThingsNetwork/lorawan-stack/blob/0c2ff932c25f454e9d24ef83322686edf2a2c83e/pkg/identityserver/gormstore/end_device.go#L228
https://github.com/TheThingsNetwork/lorawan-stack/blob/0c2ff932c25f454e9d24ef83322686edf2a2c83e/pkg/identityserver/gormstore/end_device.go#L152

The getter/setter is only called if the paths contain the top-level field `claim_authentication_code`.
The CLI only sets the sub-fields in the fieldmask (ex: `claim_authentication_code.value`). 
So ,in the registry, we check if the sub-fields are set to set the top level field.
This is the same as the implementation for the `picture` field in `L350-L352`.
(PS: This is also applicable for the bun store since it uses `ttnpb.SetFields`).

File: pkg/identityserver/end_device_registry.go (hunk 15 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go]: https://github.com/TheThingsNetwork/lorawan-stack/blob/0c2ff932c25f454e9d24ef83322686edf2a2c83e/pkg/identityserver/gormstore/end_device.go#L228
https://github.com/TheThingsNetwork/lorawan-stack/blob/0c2ff932c25f454e9d24ef83322686edf2a2c83e/pkg/identityserver/gormstore/end_device.go#L152

The getter/setter is only called if the paths contain the top-level field `claim_authentication_code`.
The CLI only sets the sub-fields in the fieldmask (ex: `claim_authentication_code.value`). 
So ,in the registry, we check if the sub-fields are set to set the top level field.
This is the same as the implementation for the `picture` field in `L350-L352`.
(PS: This is also applicable for the bun store since it uses `ttnpb.SetFields`).

File: pkg/identityserver/end_device_registry.go (hunk 14 above)
Comment (adriansmares) [reply pkg/identityserver/end_device_registry.go]: Indeed, use `ttnpb.AddFields` instead of `append` when it comes to field masks.

**Review by adriansmares — APPROVED** [review]

File: pkg/identityserver/end_device_registry.go (hunk 13 above)
Comment (nicholaspcr) [reply pkg/identityserver/end_device_registry.go:232]: Indeed the use of `ttnpb.TopLevelFields` in here might not be strictly necessary.
Considering that the values would be something like:
```
"claim_authentication_code",
"claim_authentication_code.valid_from",
"claim_authentication_code.valid_to",
"claim_authentication_code.value"
```

Doesn't the ttnpb.HasAnyField` be enough for the validation? @KrishnaIyer

File: pkg/identityserver/end_device_registry.go (hunk 18 above)
Comment (nicholaspcr) [reply pkg/identityserver/end_device_registry.go:445]: I personally prefer not allowing the creation with a `ValidTo` previous than the time of creation but at the same time it feels like an edge case where users won't end up falling in. Out of curiosity, what exactly is the test case where it would benefit from this? Wouldn't setting the `validTo` and timing out for a few seconds cover the `past valid date` case?

**Review by nicholaspcr — APPROVED** [review]
I left two small comments
1 - Regarding the usage of `ttnpb.TopLevelFields`.
2 - Not validating if `ValidTo` is before the current time.

Outside of those, this PR LGTM.

File: pkg/identityserver/end_device_registry.go (hunk 13 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go:232]: > I think HasAnyField already considers top level fields

It does not and hence the use of `ttnpb.TopLevelFields`. This is explicitly specified in the GoDoc
```
...
// Note that this function may have unexpected results when non bottom search fields are used,
// as HasAnyField([]string{"a.b"}, "a") is false.
...
func HasAnyField(requested []string, search ...string) bool
```

Without explicitly using `ttnpb.TopLevelFields`, the following command will retrieve nothing.
```
./ttn-lw-cli dev get test-app test-dev --claim-authentication-code.secret
```
whereas the following will work
```
./ttn-lw-cli dev get test-app test-dev --claim-authentication-code
```
As explained in the other comment, this is not for validation but for the fact that the gormstore implementation requires the top-level field `claim_authentication_code` to get/set any of the sub-fields. This check can be removed when we fully migrate to bun.

PS: Note that the `picture` field a few lines above works the same, as would any struct that does not allow get/set of sub fields (`gormstore` only).

File: pkg/identityserver/end_device_registry.go (hunk 18 above)
Comment (KrishnaIyer) [reply pkg/identityserver/end_device_registry.go:445]: > Wouldn't setting the validTo and timing out for a few seconds cover the past valid date case

Yes that would work. But please note that the people who use the stack are not usually engineers and who may not have the intuition to do that. Whereas simply setting something in the past date is intuitive.

File: pkg/identityserver/end_device_registry.go (hunk 13 above)
Comment (nicholaspcr) [reply pkg/identityserver/end_device_registry.go:232]: True, my bad on missing the GoDoc info. Then this looks appropriate for me.

File: pkg/identityserver/end_device_registry.go (hunk 18 above)
Comment (nicholaspcr) [reply pkg/identityserver/end_device_registry.go:445]: True enough. Then, considering that users might want to play around with the `validTo` value, this looks appropriate.
